Migrate Audience page to Inertia

Devwill2 opened the pull request:
ref #856 
### Description:
Migrates the Audience (Following) page from React-on-Rails to Inertia.js.

<h3>AI Disclosure</h3>
<p>None</p>

<hr>

<h3>Livestream Viewing Confirmation</h3>
<p>
  Have watched both the live streams end-to-end and followed all the guidelines
  asked in them.
</p>
<hr>

### Note: 
> Tabs currently use full page reloads (<a> tags) instead of SPA navigation because the UTM Links tab is not yet on Inertia. We can switch to <Link> once all analytics pages are migrated.

<h2> Video </h2>

https://github.com/user-attachments/assets/34df50ae-afee-4797-8779-9ccba2cc14d7

### Test
<img width="924" height="333" alt="Screenshot 2025-12-27 at 7 22 38 PM" src="https://github.com/user-attachments/assets/cdad0a89-42fe-43cc-b800-80d24dc1e565" />

Devwill2 commented:
@EmCousin  Currently, the tabs use regular <a> tags which cause a full page reload. This is intentional because the UTM Links page is not yet migrated to Inertia, and the shared Tab component cannot import @inertiajs/react without breaking SSR for non-Inertia pages.

Once the UTM Links page is migrated, we can simply update the Tab component to use Inertia's <Link>, enabling seamless SPA navigation across all analytics tabs
Thanks

Pradumn27 commented:
@Pradumn27 _a - will review as the assigned reviewer for the SPA issue

Pradumn27 commented on app/javascript/components/Audience/AudiencePageContent.tsx:
We also need to migrate the API endpoints to inertia-styled endpoints, so we can fully utilize inertia and use [partial-reloads](https://inertia-rails.dev/guide/partial-reloads#lazy-data-evaluation) for fetching the analytics data. Please implement that

Pradumn27 reviewed (commented):
Hey @Devwill2 _a, thanks for the PR, please check the comments.

Also the tests passing screenshot is missing in the description so please add that, and also in the video we can only see the empty state so please add data so we can better visualize what's happening.

Devwill2 commented on config/routes.rb:
```diff
@@ -768,7 +768,7 @@ def product_info_and_purchase_routes(named_routes: true)
     get "/audience" => redirect("/dashboard/audience")
     get "/dashboard/audience", to: "audience#index", as: :audience_dashboard
-    get "/audience/data/by_date/:start_time/:end_time", to: "audience#data_by_date", as: "audience_data_by_date"
+
```
This route is no longer needed because we've migrated to Inertia's partial reload pattern:

4arjun commented on app/controllers/audience_controller.rb:
```diff
@@ -1,16 +1,26 @@
 class AudienceController < Sellers::BaseController
-  before_action :set_body_id_as_app
-  before_action :set_time_range, only: %i[data_by_date]
+  before_action :set_time_range, only: %i[index]
```
We only have one symbol here.

Pradumn27 commented on app/controllers/audience_controller.rb:
```diff
@@ -50,4 +54,12 @@ def set_time_range
+        start_date: @start_date.to_date,
+        end_date: @end_date.to_date
+      )
+    end
```
why not just reuse the previous function by modifying it?

Pradumn27 commented on app/javascript/pages/Audience/Index.tsx:
```diff
@@ -1,10 +1,8 @@
-import { AudienceDataByDate, fetchAudienceDataByDate } from "$app/data/audience";
-import { AbortError } from "$app/utils/request";
-import { register } from "$app/utils/serverComponentUtil";
+import { AudienceDataByDate } from "$app/data/audience";
```
```suggestion
import { type AudienceDataByDate } from "$app/data/audience";
```

We'd like to be explicit about what we import

Devwill2 commented on app/controllers/audience_controller.rb:
```diff
@@ -1,16 +1,26 @@
+
+    if total_follower_count > 0
+      props[:audience_data] = InertiaRails.optional { fetch_audience_data }
+    end
```
```markdown
The conditional ensures we don't even define the audience_dataprop when `total_follower_count === 0`. This:
- Skips Elasticsearch query entirely for empty state
- Results in cleaner props object
- Makes it explicit to the frontend that no data is available

A lambda would still define the prop (even if lazy-evaluated), which is unnecessary overhead when we know there's no data to fetch.
```
Curious to hear what you think about this reasoning, @Pradumn27 , @EmCousin   does this make sense from your perspective?

binary-koan replied:
Personally I'd go with something like:

```ruby
render inertia: "Audience/Index", props: {
  total_follower_count: InertiaRails.always { total_follower_count },
  audience_data: -> { total_follower_count > 0 ? fetch_audience_data : nil }
}
```

* Generally we prefer explicitly passing null over omitting properties in JS
* I agree it's nice to skip the ES query entirely if we know we don't need to call it

Devwill2 commented:
Thanks @Pradumn27  Switched to lambda as suggested. Keeping the conditional since when `total_follower_count === 0`, we want to skip defining the prop entirely (not just delay the query). This avoids instantiating the analytics service and running Elasticsearch for empty states. The lambda handles lazy evaluation while the conditional handles "should this prop exist at all?"

Made the other changes: extracted `reloadAudienceData` callback, added comprehensive tests, used `ts-safe-cast`. All checks passing

Devwill2 commented:
> Thanks @Pradumn27 Switched to lambda as suggested. Keeping the conditional since when `total_follower_count === 0`, we want to skip defining the prop entirely (not just delay the query). This avoids instantiating the analytics service and running Elasticsearch for empty states. The lambda handles lazy evaluation while the conditional handles "should this prop exist at all?"
> 
> Made the other changes: extracted `reloadAudienceData` callback, added comprehensive tests, used `ts-safe-cast`. All checks passing

@binary-koan  @EmCousin, what’s your opinion on this approach?

Devwill2 commented:
@binary-koan Can you review this, please? I’ve completed all the changes and think it’s good to go.
Thanks for your time

binary-koan commented:
Yep, will check both later. In general please point to duplicates from your own PR rather than commenting on the other one, it's a bit easier to understand (e.g. "@... Could you please take a look at this? It's an alternative to #... which I see you're looking at")

Devwill2 commented:
@binary-koan I think two tests are flaky, and one of them seems related. I noticed you made some commits to simplify the spec—am I correct?
